core/silkworm/consensus/clique/clique_snapshot.cpp

```diff
@@ -0,0 +1,246 @@
+/*
+   Copyright 2020-2021 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "clique_snapshot.hpp"
+#include <silkworm/common/endian.hpp>
+#include <silkworm/common/util.hpp>
+#include <silkworm/crypto/ecdsa.hpp>
+#include <cstring>
+
+namespace silkworm {
+
+std::array<uint8_t, 8> kNonceAuthorize =   {0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff};
+std::array<uint8_t, 8> kNonceUnauthorize = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+//! \brief Updated snapshot by adding headers
+//! \param headers: list of headers to add.
+ValidationResult CliqueSnapshot::add_header(const BlockHeader& header, const evmc::address& signer, const CliqueConfig& config) {
+    auto hash{header.hash()};
+    auto tmp_recents{recents_}; // We only modify snapshot after checks are done.
+    // Delete the oldest signer from the recent list to allow it signing again
+    if (recents_.size() > signers_.size() / 2) {
+        tmp_recents.pop_back();
+    }
+
+    if (std::find(signers_.begin(), signers_.end(), signer) == signers_.end()) {
+        return ValidationResult::kUnauthorizedSigner;
+    }
+
+    // Remove any votes on checkpoint blocks
+    if (header.number % config.epoch == 0) {
+        tallies_.clear();
+    }
+
+    if (std::find(tmp_recents.begin(), tmp_recents.end(), signer) != tmp_recents.end()) {
+        return ValidationResult::kRecentlySigned;
+    }
+
+    // We check what the vote is
+    bool authorize;
+    if (header.nonce == kNonceAuthorize) {
+        authorize = true;
+    } else if (header.nonce == kNonceUnauthorize) {
+        authorize = false;
+    } else {
+        return ValidationResult::kInvalidVote;
+    }
+    
+    tmp_recents.push_front(signer);
+    recents_ = tmp_recents;
```

Comment:
Besides the fact this should not be here this is dramatically expensive.
Should you need to move contents of tmp_recents to recents_ (and eventually discard tmp_recents_) use 

`recents_.swap(tmp_recents);`

which only moves pointers to their respective data.